internal/app/machined/pkg/controllers/config/node_labels.go

```diff
@@ -0,0 +1,334 @@
+	k8sLabels, err := k8sClient.GetNodeLabels(ctx, nodeName)
+	if err != nil {
+		return fmt.Errorf("error getting node labels from kubernetes: %w", err)
+	}
+
+	stateNodeLabels, err := getStateNodeLabels(ctx, r)
+	if err != nil {
+		return err
+	}
+
+	for _, stateLabel := range stateNodeLabels.Items {
+		stateLabelSpec := stateLabel.(*config.NodeLabel).TypedSpec()
+
+		k8sLabelValue, ok := k8sLabels[stateLabelSpec.Key]
+		if ok {
+			delete(k8sLabels, stateLabelSpec.Key)
+
+			if k8sLabelValue != stateLabelSpec.Value {
+				if err := updateLabel(ctx, r, stateLabel, k8sLabelValue, false); err != nil {
+					return fmt.Errorf("error updating node label resource: %w", err)
+				}
+			}
+		}
+	}
+
+	return addLabelsToState(ctx, r, k8sLabels, mcNodeLabels)
+}
+
+func addLabelsToState(ctx context.Context, r controller.Runtime, k8sLabels, mcNodeLabels map[string]string) error {
+	for key, value := range k8sLabels {
+		createdByTalos := false
+
+		mcValue, ok := mcNodeLabels[key]
+		if ok && mcValue == value {
+			createdByTalos = true
+		}
+
+		labelRes := config.NewPopulatedNodeLabel(key, key, value, createdByTalos)
+
+		if err := r.Create(ctx, labelRes); err != nil {
+			return fmt.Errorf("failed to create node label resource: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func copyMap(src map[string]string) map[string]string {
+	dst := map[string]string{}
+
+	for key, value := range src {
+		dst[key] = value
+	}
+
+	return dst
+}
+
+func (ctrl *NodeLabelsController) computeChangesToMake(stateNodeLabels *resource.List, mcNodeLabels talosconfig.NodeLabels) (map[string]struct{}, map[string]string, map[string]string) {
+	labelIDsToDelete := map[resource.ID]struct{}{}
+
+	mcNodeLabelsMap := map[string]string{}
+	labelsToAdd := map[string]string{}
+	labelsToChange := map[string]string{}
+
+	for key, value := range mcNodeLabels {
+		mcNodeLabelsMap[key] = value
+		labelsToAdd[key] = value
+	}
+
+	for _, stateLabel := range stateNodeLabels.Items {
+		stateLabelSpec := stateLabel.(*config.NodeLabel).TypedSpec()
+
+		mcLabelValue, ok := mcNodeLabelsMap[stateLabelSpec.Key]
+		if ok {
+			delete(labelsToAdd, stateLabelSpec.Key)
+
+			if mcLabelValue != stateLabelSpec.Value {
+				labelsToChange[stateLabelSpec.Key] = mcLabelValue
+			}
+		} else if stateLabelSpec.CreatedByTalos {
+			labelIDsToDelete[stateLabel.Metadata().ID()] = struct{}{}
+		}
+	}
+
+	return labelIDsToDelete, labelsToAdd, labelsToChange
+}
+
+func (ctrl *NodeLabelsController) reconcileNodeLabels(
+	ctx context.Context,
+	r controller.Runtime,
+	nodeName string,
+	mcNodeLabels talosconfig.NodeLabels,
+	k8sClient NodeLabelsK8sClientIface,
+	logger *zap.Logger,
+) error {
+	stateNodeLabels, err := getStateNodeLabels(ctx, r)
+	if err != nil {
+		return err
+	}
+
+	labelIDsToDelete, labelsToAdd, labelsToChange := ctrl.computeChangesToMake(stateNodeLabels, mcNodeLabels)
+
+	// We are making a compromise here.
```

Comment:
we don't need to make a compromise here necessarily

the problem of figuring out "ownership" is more or less already solved in Talos e.g. for network addresses

the supposed way is to implement a proper finalizer flow on "desired" label resources - this controller should set its own finalizer asap, and then before letting the label resource to be torn down (when the label is removed from the config), the controller will make sure the label is removed from the Kubernetes node

in other words, ownership is determined by configuration. 

with the node labels removal is not trivial though, as Kubernetes might be down (while Linux kernel can't be for addresses), so controller might not be able to remove the label from the node, while after a reboot Talos will lose track of what labels need to be removed

so we either say we don't support label removal at all, or we say that label removal is not fully guaranteed, it works 99% of the time (if Kubernetes is up)